hack/e2e.go

```diff
@@ -332,19 +347,217 @@ func Build() error {
+	}
+	zones := strings.Split(getenvDefault("KOPS_AWS_ZONES", "us-west-2a"), ",")
+	nodesStr := getenvDefault("KOPS_AWS_NODES", "2")
+	nodes, err := strconv.Atoi(nodesStr)
+	if err != nil {
+		return nil, fmt.Errorf("KOPS_AWS_NODES set to invalid integer %q", nodesStr)
+	}
+	cluster := os.Getenv("KOPS_CLUSTER_NAME")
+	if cluster == "" {
+		return nil, fmt.Errorf("KOPS_CLUSTER_NAME must be set for kops deployment")
+	}
+	f, err := ioutil.TempFile("", "kops-kubecfg")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	kubecfg := f.Name()
+	if err := f.Chmod(0600); err != nil {
+		return nil, err
+	}
+	if err := os.Setenv("KUBECONFIG", kubecfg); err != nil {
+		return nil, err
+	}
+	// Set KUBERNETES_CONFORMANCE_TEST so the auth info is picked up
+	// from kubectl instead of bash inference.
+	if err := os.Setenv("KUBERNETES_CONFORMANCE_TEST", "yes"); err != nil {
+		return nil, err
+	}
+	// Set KUBERNETES_CONFORMANCE_PROVIDER to override the
+	// cloudprovider for KUBERNETES_CONFORMANCE_TEST.
+	if err := os.Setenv("KUBERNETES_CONFORMANCE_PROVIDER", "aws"); err != nil {
+		return nil, err
+	}
+	// ZONE is required by the AWS e2e tests
+	if err := os.Setenv("ZONE", zones[0]); err != nil {
+		return nil, err
+	}
+	return &kops{
+		binary:    binary,
+		tmpBinary: tmpBinary,
+		zones:     zones,
+		nodes:     nodes,
+		cluster:   cluster,
+		kubecfg:   kubecfg,
+	}, nil
+}
+
+func (k kops) Up() error {
+	if err := finishRunning("kops config", exec.Command(
+		k.binary, "create", "cluster",
+		"--name", k.cluster,
+		"--node-count", strconv.Itoa(k.nodes),
+		"--zones", strings.Join(k.zones, ","))); err != nil {
+		return fmt.Errorf("kops configuration failed: %v", err)
+	}
+	if err := finishRunning("kops update", exec.Command(k.binary, "update", "cluster", k.cluster, "--yes")); err != nil {
+		return fmt.Errorf("kops bringup failed: %v", err)
+	}
+	// TODO(zmerlynn): More cluster validation. This should perhaps be
+	// added to kops and not here, but this is a fine place to loop
+	// for now.
+	for stop := time.Now().Add(10 * time.Minute); time.Now().Before(stop); time.Sleep(30 * time.Second) {
+		n, err := k.ClusterSize()
+		if err != nil {
+			log.Printf("Can't get cluster size, sleeping: %v", err)
+			continue
+		}
+		if n < k.nodes {
+			log.Printf("%d (current nodes) < %d (requested), sleeping", n, k.nodes)
+			continue
+		}
+		return nil
+	}
+	return fmt.Errorf("kops bringup timed out")
+}
+
+func (k kops) IsUp() bool {
+	n, err := k.ClusterSize()
+	if err != nil {
+		log.Printf("IsUp: false: %v", err)
+		return false
+	}
+	log.Printf("IsUp: %v (%d > 0)", n > 0, n)
+	return n > 0
+}
+
+func (k kops) SetupKubecfg() error {
+	info, err := os.Stat(k.kubecfg)
+	if err != nil {
+		return err
+	}
+	if info.Size() > 0 {
+		// Assume that if we already have it, it's good.
+		return nil
+	}
+	if err := finishRunning("kops export", exec.Command(k.binary, "export", "kubecfg", k.cluster)); err != nil {
```

Comment:
Also, in case the `kubecfg` section is unclear, this is striving to ensure that if you point it at a specific cluster, it tests that cluster, hence the `/tmp` kubecfg that's basically isolated to the cluster in question. There's a longstanding badness in the GCE bash-isms where it's just implicitly adding it to the default kubecfg, which can get sliced if you run `-up -test` in parallel, e.g.